<a href="https://colab.research.google.com/github/uroy98/CapitalOneProject/blob/main/CapitalOneData.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Will delete based on need. Few trials and errors should be there

In [ ]:
!git clone https://github.com/CapitalOneRecruiting/DS


Cloning into 'DS'...
remote: Enumerating objects: 4, done.
remote: Total 4 (delta 0), reused 0 (delta 0), pack-reused 4 (from 1)
Receiving objects: 100% (4/4), 29.06 MiB | 13.92 MiB/s, done.


In [ ]:
import pandas as pd
import requests
import io

# Correct URL pointing to the raw zip file
url = "https://raw.githubusercontent.com/CapitalOneRecruiting/DS/master/transactions.zip"

# Download the zip file content
response = requests.get(url)
response.raise_for_status()  # Raise an exception for bad responses (4xx or 5xx)

# Read the zip file into a pandas DataFrame
df = pd.read_csv(io.BytesIO(response.content), compression='zip')
print(df.head())

   {"accountNumber": "737265056"   "customerId": "737265056"  \
0  {"accountNumber": "737265056"   "customerId": "737265056"   
1  {"accountNumber": "737265056"   "customerId": "737265056"   
2  {"accountNumber": "737265056"   "customerId": "737265056"   
3  {"accountNumber": "830329091"   "customerId": "830329091"   
4  {"accountNumber": "830329091"   "customerId": "830329091"   

    "creditLimit": 5000.0   "availableMoney": 5000.0  \
0   "creditLimit": 5000.0   "availableMoney": 5000.0   
1   "creditLimit": 5000.0   "availableMoney": 5000.0   
2   "creditLimit": 5000.0   "availableMoney": 5000.0   
3   "creditLimit": 5000.0   "availableMoney": 5000.0   
4   "creditLimit": 5000.0   "availableMoney": 5000.0   

    "transactionDateTime": "2016-08-13T14:27:32"   "transactionAmount": 98.55  \
0   "transactionDateTime": "2016-10-11T05:05:54"   "transactionAmount": 74.51   
1   "transactionDateTime": "2016-11-08T09:18:39"    "transactionAmount": 7.47   
2   "transactionDateTime": "2016-12

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 786362 entries, 0 to 786361
Data columns (total 29 columns):
 #   Column                                         Non-Null Count   Dtype 
---  ------                                         --------------   ----- 
 0   {"accountNumber": "737265056"                  786362 non-null  object
 1    "customerId": "737265056"                     786362 non-null  object
 2    "creditLimit": 5000.0                         786362 non-null  object
 3    "availableMoney": 5000.0                      786362 non-null  object
 4    "transactionDateTime": "2016-08-13T14:27:32"  786362 non-null  object
 5    "transactionAmount": 98.55                    786362 non-null  object
 6    "merchantName": "Uber"                        786362 non-null  object
 7    "acqCountry": "US"                            786362 non-null  object
 8    "merchantCountryCode": "US"                   786362 non-null  object
 9    "posEntryMode": "02"                          7

In [ ]:
df.head()

{"accountNumber": "737265056"   "customerId": "737265056"  \
0  {"accountNumber": "737265056"   "customerId": "737265056"   
1  {"accountNumber": "737265056"   "customerId": "737265056"   
2  {"accountNumber": "737265056"   "customerId": "737265056"   
3  {"accountNumber": "830329091"   "customerId": "830329091"   
4  {"accountNumber": "830329091"   "customerId": "830329091"   

    "creditLimit": 5000.0   "availableMoney": 5000.0  \
0   "creditLimit": 5000.0   "availableMoney": 5000.0   
1   "creditLimit": 5000.0   "availableMoney": 5000.0   
2   "creditLimit": 5000.0   "availableMoney": 5000.0   
3   "creditLimit": 5000.0   "availableMoney": 5000.0   
4   "creditLimit": 5000.0   "availableMoney": 5000.0   

    "transactionDateTime": "2016-08-13T14:27:32"   "transactionAmount": 98.55  \
0   "transactionDateTime": "2016-10-11T05:05:54"   "transactionAmount": 74.51   
1   "transactionDateTime": "2016-11-08T09:18:39"    "transactionAmount": 7.47   
2   "transactionDateTime": "2016-12-10T02:14:50"    "transactionAmount": 7.47   
3   "transactionDateTime": "2016-03-24T21:04:46"   "transactionAmount": 71.18   
4   "transactionDateTime": "2016-04-19T16:24:27"   "transactionAmount": 30.76   

                   "merchantName": "Uber"   "acqCountry": "US"  \
0           "merchantName": "AMC #191138"   "acqCountry": "US"   
1            "merchantName": "Play Store"   "acqCountry": "US"   
2            "merchantName": "Play Store"   "acqCountry": "US"   
3   "merchantName": "Tim Hortons #947751"   "acqCountry": "US"   
4      "merchantName": "In-N-Out #422833"   "acqCountry": "US"   

    "merchantCountryCode": "US"   "posEntryMode": "02"  ...  \
0   "merchantCountryCode": "US"   "posEntryMode": "09"  ...   
1   "merchantCountryCode": "US"   "posEntryMode": "09"  ...   
2   "merchantCountryCode": "US"   "posEntryMode": "09"  ...   
3   "merchantCountryCode": "US"   "posEntryMode": "02"  ...   
4   "merchantCountryCode": "US"   "posEntryMode": "02"  ...   

                 "echoBuffer": ""   "currentBalance": 0.0  \
0   "transactionType": "PURCHASE"        "echoBuffer": ""   
1                "echoBuffer": ""   "currentBalance": 0.0   
2                "echoBuffer": ""   "currentBalance": 0.0   
3   "transactionType": "PURCHASE"        "echoBuffer": ""   
4   "transactionType": "PURCHASE"        "echoBuffer": ""   

       "merchantCity": ""   "merchantState": ""     "merchantZip": ""  \
0   "currentBalance": 0.0    "merchantCity": ""   "merchantState": ""   
1      "merchantCity": ""   "merchantState": ""     "merchantZip": ""   
2      "merchantCity": ""   "merchantState": ""     "merchantZip": ""   
3   "currentBalance": 0.0    "merchantCity": ""   "merchantState": ""   
4   "currentBalance": 0.0    "merchantCity": ""   "merchantState": ""   

    "cardPresent": false   "posOnPremises": ""   "recurringAuthInd": ""  \
0      "merchantZip": ""   "posOnPremises": ""   "recurringAuthInd": ""   
1   "cardPresent": false   "posOnPremises": ""   "recurringAuthInd": ""   
2   "cardPresent": false   "posOnPremises": ""   "recurringAuthInd": ""   
3      "merchantZip": ""   "posOnPremises": ""   "recurringAuthInd": ""   
4      "merchantZip": ""   "posOnPremises": ""   "recurringAuthInd": ""   

    "expirationDateKeyInMatch": false   "isFraud": false}  
0   "expirationDateKeyInMatch": false   "isFraud": false}  
1   "expirationDateKeyInMatch": false   "isFraud": false}  
2   "expirationDateKeyInMatch": false   "isFraud": false}  
3   "expirationDateKeyInMatch": false   "isFraud": false}  
4   "expirationDateKeyInMatch": false   "isFraud": false}  

[5 rows x 29 columns]

Since the data is in line-delimited JSON format format, we can use Python with Pandas and the JSON library to parse and convert it into a structured DataFrame.

# Question 1: Load

* Programmatically download and load into your favorite analytical tool the transactions data. This data, which is in line-delimited JSON format, can be found in the Github link - Done

In [ ]:
import json
import pandas as pd
import requests
import zipfile
import io
import os

# Define the URL of the GitHub ZIP file.
# Replace 'username' and 'repository' with the actual repository details.
zip_url = "https://raw.githubusercontent.com/CapitalOneRecruiting/DS/master/transactions.zip"

# Download the ZIP file from GitHub.
print("Downloading ZIP file from GitHub...")
response = requests.get(zip_url)
response.raise_for_status()  # Ensure that an HTTP error is raised for unsuccessful status codes.

# Unzip the downloaded file into a designated local directory.
extraction_dir = "/content"
print("Extracting the ZIP file...")
with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    z.extractall(extraction_dir)

# Define the path to the text file containing JSON lines.
# Adjust the file path as per the repository structure. For example, if the file is located in
# a folder named 'repository-main' (the default folder name after extraction) and is called 'transactions.txt'.
transactions_file_path = os.path.join("/content/transactions.txt")

# Open the text file and read all the JSON lines.
print("Reading the transactions file...")
with open(transactions_file_path, 'r') as file:
    json_lines = file.readlines()

# Parse each non-empty line into a Python dictionary.
print("Parsing JSON lines...")
data = [json.loads(line.strip()) for line in json_lines if line.strip()]

# Assuming each record has a uniform structure, extract the field names from the first record.
field_names = list(data[0].keys())

print("Field Names:")
print(field_names)

# Convert the list of dictionaries into a pandas DataFrame for further analysis.
df = pd.DataFrame(data)
print("\nDataFrame:")
print(df)


Extracting the ZIP file...
Reading the transactions file...
Parsing JSON lines...
Field Names:
['accountNumber', 'customerId', 'creditLimit', 'availableMoney', 'transactionDateTime', 'transactionAmount', 'merchantName', 'acqCountry', 'merchantCountryCode', 'posEntryMode', 'posConditionCode', 'merchantCategoryCode', 'currentExpDate', 'accountOpenDate', 'dateOfLastAddressChange', 'cardCVV', 'enteredCVV', 'cardLast4Digits', 'transactionType', 'echoBuffer', 'currentBalance', 'merchantCity', 'merchantState', 'merchantZip', 'cardPresent', 'posOnPremises', 'recurringAuthInd', 'expirationDateKeyInMatch', 'isFraud']

DataFrame:
       accountNumber customerId  creditLimit  availableMoney  \
0          737265056  737265056       5000.0         5000.00   
1          737265056  737265056       5000.0         5000.00   
2          737265056  737265056       5000.0         5000.00   
3          737265056  737265056       5000.0         5000.00   
4          830329091  830329091       5000.0         

In [ ]:
df.head()

accountNumber customerId  creditLimit  availableMoney  transactionDateTime  \
0     737265056  737265056       5000.0          5000.0  2016-08-13T14:27:32   
1     737265056  737265056       5000.0          5000.0  2016-10-11T05:05:54   
2     737265056  737265056       5000.0          5000.0  2016-11-08T09:18:39   
3     737265056  737265056       5000.0          5000.0  2016-12-10T02:14:50   
4     830329091  830329091       5000.0          5000.0  2016-03-24T21:04:46   

   transactionAmount         merchantName acqCountry merchantCountryCode  \
0              98.55                 Uber         US                  US   
1              74.51          AMC #191138         US                  US   
2               7.47           Play Store         US                  US   
3               7.47           Play Store         US                  US   
4              71.18  Tim Hortons #947751         US                  US   

  posEntryMode  ... echoBuffer currentBalance merchantCity merchantState  \
0           02  ...                       0.0                              
1           09  ...                       0.0                              
2           09  ...                       0.0                              
3           09  ...                       0.0                              
4           02  ...                       0.0                              

  merchantZip cardPresent posOnPremises recurringAuthInd  \
0                   False                                  
1                    True                                  
2                   False                                  
3                   False                                  
4                    True                                  

  expirationDateKeyInMatch isFraud  
0                    False   False  
1                    False   False  
2                    False   False  
3                    False   False  
4                    False   False  

[5 rows x 29 columns]

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 786363 entries, 0 to 786362
Data columns (total 29 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   accountNumber             786363 non-null  object 
 1   customerId                786363 non-null  object 
 2   creditLimit               786363 non-null  float64
 3   availableMoney            786363 non-null  float64
 4   transactionDateTime       786363 non-null  object 
 5   transactionAmount         786363 non-null  float64
 6   merchantName              786363 non-null  object 
 7   acqCountry                786363 non-null  object 
 8   merchantCountryCode       786363 non-null  object 
 9   posEntryMode              786363 non-null  object 
 10  posConditionCode          786363 non-null  object 
 11  merchantCategoryCode      786363 non-null  object 
 12  currentExpDate            786363 non-null  object 
 13  accountOpenDate           786363 non-null  o

# Question 1: Load

* Please describe the structure of the data. Number of records and fields in each record?

**Answer**: Since the data is in line-delimited JSON format format, I have used a Python code that demonstrates how to download the "transactions.zip" file from GitHub, extract its contents, and then process a text file containing newline-delimited JSON objects to separate the field names from the data.
The given dataset has **786363 records** and **29 fields** in each record.

# Question 1:Load
* Please provide some additional basic summary statistics for each field. Be sure to include a count of null, minimum, maximum, and unique values where appropriate.